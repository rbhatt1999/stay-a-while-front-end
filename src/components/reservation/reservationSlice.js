import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import axios from 'axios';

const LIST_RESERVATIONS = 'stayAWhile/list_reservations';
const DELETE_RESERVATION = 'stayAWhile/delete_reservation';
const baseUrl = 'http://localhost:3000/reservations';

export const getListReservations = createAsyncThunk(LIST_RESERVATIONS, async (token) => {
  try {
    const options = {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    };
    const response = await axios.request(baseUrl, options);
    const responseData = response.data;
    if (responseData.error) {
      throw new Error(responseData.error);
    } else {
      const reservationList = responseData.data.reservation;
      return reservationList;
    }
  } catch (error) {
    return { error: error.message };
  }
});

export const deleteReservation = createAsyncThunk(DELETE_RESERVATION,
  async (payload) => {
    const deleteUrl = `${baseUrl}/${payload.id}`;
    try {
      const options = {
        headers: {
          Authorization: `Bearer ${payload.token}`,
        },
      };
      const response = await axios.delete(deleteUrl, options);
      const responseData = response.data;
      if (response.error) {
        throw new Error(responseData.error);
      } else {
        return responseData;
      }
    } catch (error) {
      return { error: error.message };
    }
  });

const reservationSlice = createSlice({
  name: 'reservations',
  initialState: {
    reservations: [],
    loading: false,
    error: null,
  },
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(getListReservations.pending, (state) => {
        const newState = { ...state, loading: true };
        return newState;
      })
      .addCase(getListReservations.fulfilled, (state, action) => {
        const newState = { ...state, loading: false, reservations: action.payload };
        return newState;
      })
      .addCase(getListReservations.rejected, (state, action) => {
        const newState = { ...state, loading: false, error: action.error };
        return newState;
      })
      .addCase(deleteReservation.pending, (state) => {
        const newState = { ...state, loading: true };
        return newState;
      })
      .addCase(deleteReservation.fulfilled, (state, action) => {
        const filteredReservations = state.reservations.filter(
          (reservation) => reservation.id !== action.payload.id,
        );
        const newState = { ...state, loading: false, reservations: filteredReservations };
        return newState;
      })
      .addCase(deleteReservation.rejected, (state, action) => {
        const newState = { ...state, loading: false, error: action.error };
        return newState;
      });
  },
});
export default reservationSlice.reducer;
