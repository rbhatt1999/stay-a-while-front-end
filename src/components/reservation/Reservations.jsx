import React, { useEffect, useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import ReactPaginate from 'react-paginate';
import { deleteReservation, getListReservations } from './reservationSlice';
import styles from './Reservations.module.css';
import { fetchHomeStays } from '../home/HomeSlice';

export default function Reservation() {
  const dispatch = useDispatch();
  const token = localStorage.getItem('token');
  const { reservations } = useSelector((state) => state.reservation);
  const homeStayList = useSelector((state) => state.home.listings);
  const [currentPage, setCurrentPage] = useState(0);
  const [totalPages, setTotalPages] = useState(0);
  const itemsPerPage = 6;
  const startIndex = currentPage * itemsPerPage;
  const endIndex = startIndex + itemsPerPage;
  const handlePageChange = (selectedPage) => {
    setCurrentPage(selectedPage.selected);
  };

  useEffect(() => {
    if (reservations) {
      setTotalPages(Math.ceil(reservations.length / itemsPerPage));
    }
  }, [reservations]);

  useEffect(() => {
    dispatch(getListReservations(token));
    dispatch(fetchHomeStays(token));
  }, [dispatch, token]);

  const handleDelete = async (reservationId) => {
    const payload = {
      id: reservationId,
      token,
    };
    dispatch(deleteReservation(payload));
  };

  const subset = reservations.slice(startIndex, endIndex);

  return (
    <div className={styles.container}>
      {reservations.length === 0 ? (
        <div className="flex flex-col items-center justify-center">
          <h1 className="text-center text-gray-500">
            You have no reservations yet.
          </h1>
          <h2 className="text-center text-gray-500">
            Go to the home page and book a home.
          </h2>
        </div>
      ) : (
        <div className={styles.card_container}>
          {subset.map((reservation) => {
            const homeStay = homeStayList.find(
              (h) => h.id === reservation.home_stay_id,
            );
            if (!homeStay) return null;
            return (
              <div key={reservation.id}>
                <div className={styles.card}>
                  <div className={styles.card__img}>
                    <img
                      src={
                        homeStay.images
                        && homeStay.images.length > 0
                        && homeStay.images[0].url
                      }
                      alt={homeStay.name}
                    />
                  </div>
                  <div className={styles.card__descr_wrapper}>
                    <h2 className={styles.card__title}>{homeStay.name}</h2>
                    <p className={styles.price}>
                      <span className={styles.total_price}>
                        $
                        {(homeStay.price
                          * (new Date(reservation.end_date).getTime()
                            - new Date(reservation.start_date).getTime()))
                          / (1000 * 3600 * 24)}
                      </span>
                      {' '}
                      |
                      <span className={styles.price_rate}>
                        $
                        {homeStay.price}
                        {' '}
                        per day
                      </span>
                    </p>
                    <p className={styles.card__descr}>
                      <span>
                        {reservation.start_date}
                        {' '}
                        -
                        {' '}
                        {reservation.end_date}
                      </span>
                    </p>
                    <button
                      type="button"
                      className="btn btn-danger px-4 py-2 rounded-lg"
                      onClick={() => handleDelete(reservation.id)}
                    >
                      Cancel
                    </button>
                  </div>
                </div>
              </div>
            );
          })}
        </div>
      )}
      <ReactPaginate
        itemClass="pageitem"
        linkClass="pagelink"
        previousLabel="Prev"
        nextLabel="Next"
        pageCount={totalPages}
        onPageChange={handlePageChange}
        containerClassName="pagination justify-content-center"
        pageClassName="pageitem"
        pageLinkClassName="pagelink"
        activeClassName="active"
        previousClassName="pageitem"
        previousLinkClassName="pagelink"
        nextClassName="pageitem"
        nextLinkClassName="pagelink"
      />
    </div>
  );
}
