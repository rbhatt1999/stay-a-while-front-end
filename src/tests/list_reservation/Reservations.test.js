import React from 'react';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import { Provider, useSelector } from 'react-redux';
import { render, act, screen } from '@testing-library/react';
import { configureStore } from '@reduxjs/toolkit';
import axios from 'axios';
import reservationSlice, { getListReservations } from '../../components/reservation/reservationSlice';
import Reservations from '../../components/reservation/Reservations';
import reservations from './reservations.json';
import homeSlice, { fetchHomeStays } from '../../components/home/HomeSlice';
import homestays from '../delete_homeStay/home_stays.json';

jest.mock('axios');
jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
}));

const store = configureStore({
  reducer: {
    reservation: reservationSlice,
    home: homeSlice.reducer,
  },
});

describe('Reservations', () => {
  beforeEach(async () => {
    axios.get.mockResolvedValue({
      data: {
        data: {
          reservation: reservations.data.reservation,
        },
      },
    });
    useSelector.mockImplementation((selectorFn) => selectorFn({
      reservation: { reservations: reservations.data.reservation },
      home: { listings: homestays.data.home_stays },
    }));

    await act(() => {
      store.dispatch(getListReservations());
      store.dispatch(fetchHomeStays());
    });

    await act(async () => {
      render(
        <Provider store={store}>
          <Reservations />
        </Provider>,
      );
    });
  });

  test('should have a submit button', () => {
    expect(screen.getAllByRole('button', { name: /Cancel/i })).toHaveLength(2);
  });

  test('should match screenshot', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Reservations />
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('it should have an image of home stay', () => {
    expect(screen.getAllByRole('img')).toHaveLength(2);
  });
});
