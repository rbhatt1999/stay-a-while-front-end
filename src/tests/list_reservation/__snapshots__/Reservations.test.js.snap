// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reservations should match screenshot 1`] = `
<div
  className="container"
>
  <div
    className="card_container"
  >
    <div>
      <div
        className="card"
      >
        <div
          className="card__img"
        >
          <img
            alt="Cottage by Maxwell"
            src="https://images.pexels.com/photos/2102587/pexels-photo-2102587.jpeg?auto=compress&cs=tinysrgb&w=600"
          />
        </div>
        <div
          className="card__descr_wrapper"
        >
          <h2
            className="card__title"
          >
            Cottage by Maxwell
          </h2>
          <p
            className="price"
          >
            <span
              className="total_price"
            >
              $
              1200
            </span>
             
            |
            <span
              className="price_rate"
            >
              $
              150
               
              per day
            </span>
          </p>
          <p
            className="card__descr"
          >
            <span>
              2023-05-18
               
              -
               
              2023-05-26
            </span>
          </p>
          <button
            className="btn btn-danger px-4 py-2 rounded-lg"
            onClick={[Function]}
            type="button"
          >
            Cancel
          </button>
        </div>
      </div>
    </div>
    <div>
      <div
        className="card"
      >
        <div
          className="card__img"
        >
          <img
            alt="Home by Sam"
            src="https://images.pexels.com/photos/259588/pexels-photo-259588.jpeg?auto=compress&cs=tinysrgb&w=600"
          />
        </div>
        <div
          className="card__descr_wrapper"
        >
          <h2
            className="card__title"
          >
            Home by Sam
          </h2>
          <p
            className="price"
          >
            <span
              className="total_price"
            >
              $
              1575
            </span>
             
            |
            <span
              className="price_rate"
            >
              $
              175
               
              per day
            </span>
          </p>
          <p
            className="card__descr"
          >
            <span>
              2023-05-16
               
              -
               
              2023-05-25
            </span>
          </p>
          <button
            className="btn btn-danger px-4 py-2 rounded-lg"
            onClick={[Function]}
            type="button"
          >
            Cancel
          </button>
        </div>
      </div>
    </div>
  </div>
  <ul
    aria-label="Pagination"
    className="pagination justify-content-center"
    role="navigation"
  >
    <li
      className="pageitem disabled"
    >
      <a
        aria-disabled="true"
        aria-label="Previous page"
        className="pagelink "
        onClick={[Function]}
        onKeyPress={[Function]}
        rel="prev"
        role="button"
        tabIndex="-1"
      >
        Prev
      </a>
    </li>
    <li
      className="pageitem"
    >
      <a
        aria-disabled="false"
        aria-label="Next page"
        className="pagelink"
        onClick={[Function]}
        onKeyPress={[Function]}
        rel="next"
        role="button"
        tabIndex="0"
      >
        Next
      </a>
    </li>
  </ul>
</div>
`;
